### ISSUE TITLE
Loops implementation?

### ISSUE THREAD

**Issue by blza** [jdbcx/jdbcx#42]
Is or will there be a way to generate and run queries in a loop consecutively or in parallel using jdbcx? Say I'd like to use it to invoke consecutive drop of partitions in Vertica DB using something like this:
```sql
{% for i in range(0,255): %}
select DROP_PARTITIONS('DDS.A_Some_Table', {{ script: ${i} }}, {{ script: ${i} }});
```

**Comment by zhicwu** [comment]
Hi @blza, unlike a real template language, looping is implemented in a unique way in JDBCX. Please refer to examples below:

```sql
-- issue 4 x 5 = 20 'select <n> num_col, <c> str_col' queries
select {{ script: [1,2,3,4] }} num_col, '{{ script: ["a","b","c","d","e"] }}' str_col

-- issue 10 'select <n> as `col<n>`' queries, only one loop for same blocks(in v0.4+)
select {{ script: Array(10).fill().map((element, index) => index + 1) }}
  as `col{{ script: Array(10).fill().map((element, index) => index + 1) }}`

-- same as above, but use SQL query(slower) instead of scripting to generate numbers
select {{ db.ch-play: select * from numbers(1, 10) }}
```

More examples using docker and DuckDB:
```bash
# get the latest unreleased v0.4.0
docker pull jdbcx/jdbcx

# issue two queries ('select 1' and 'select 2') one by one and return consolidated results
docker run --rm -it -e DRIVER_OPTS="-Dverbose=true -DnoProperties=true" jdbcx/jdbcx 'jdbcx:duckdb:' \
  'select {{script:[1,2]}}'

# create a table for testing
docker run --rm -it -e DRIVER_OPTS="-Dverbose=true -DnoProperties=true" -v `pwd`/duckdb:/data \
  jdbcx/jdbcx 'jdbcx:duckdb:/data/test' 'create table a(num INTEGER, str VARCHAR)'

# insert 3 x 2 = 6 rows
docker run --rm -it -e DRIVER_OPTS="-Dverbose=true -DnoProperties=true" -v `pwd`/duckdb:/data \
  jdbcx/jdbcx 'jdbcx:duckdb:/data/test' "insert into a values({{script:[1,2,3]}}, '{{script:['a','b']}}')"

# add 3 nullable varchar columns
docker run --rm -it -e DRIVER_OPTS="-Dverbose=true -DnoProperties=true" -v `pwd`/duckdb:/data \
  jdbcx/jdbcx 'jdbcx:duckdb:/data/test' '{%var:n=col%}alter table a add column {{script:${n}}}{{script:[1,2,3]}} varchar'

# query the table (6 rows, 5 columns)
docker run --rm -it -e DRIVER_OPTS="-Dverbose=true -DnoProperties=true" -v `pwd`/duckdb:/data \
  jdbcx/jdbcx 'jdbcx:duckdb:/data/test' 'select * from a'
```

**Comment by blza** [comment]
Thanks a lot, an interesting way, will try to figure it out. Btw `examples` link in the README leads to some folder with no examples present (namely https://github.com/jdbcx/jdbcx/tree/main/docker/app/.jdbcx/).

**Comment by blza** [comment]
And more for the very first example I get the error:

```
SQL Error [HY000]: No script language detected in classpath(loader=io.github.jdbcx.ExpandedUrlClassLoader@78c51ac7(parent=org.jkiss.dbeaver.registry.driver.DriverClassLoader@7bfaa86urls=[/Users/alosev/Library/DBeaverData/drivers/maven/maven-central/com.vertica.jdbc/])).
```

I'm using dbeaver jdbcx driver and set classpath for Vertica by setting `jdbcx.custom.classpath` driver property to `"/Users/<my_user>/Library/DBeaverData/drivers/maven/maven-central/com.vertica.jdbc/"` (it successfully connects to Vertica and is able to run queries against DB).

**Comment by blza** [comment]
This maybe related to issue mentioned in README `Scripting does not work on DBeaver` and the solution `use JDK instead of JRE`. Where should I change JRE to JDK?

**Comment by zhicwu** [comment]
Sorry for the unclear documents. It's still in the early stages, so please bear with the tedious steps required to get things working ;)

> Btw examples link in the README leads to some folder with no examples present (namely https://github.com/jdbcx/jdbcx/tree/main/docker/app/.jdbcx/)

It's intentional :) It's highly recommended to have a similar configuration([config.properties](https://github.com/jdbcx/jdbcx/tree/main/docker/app/.jdbcx/config.properties)) in your home directory(`<user home>/.jdbcx`), so that you don't have to specify driver settings like `jdbcx.custom.classpath` and share the same configuration across all JDBCX connections defined in DBeaver or any other IDEs.

> SQL Error [HY000]: No script language detected in classpath

JDBCX driver was added in DBeaver in 23.2.1, and the scripting issue was fixed since 23.2.2, so you don't have to use JDK anymore. You just need to download and add below two jars in your classpath(or put the jars into the directory specified by `jdbcx.custom.classpath`):
* https://repo1.maven.org/maven2/org/mozilla/rhino/1.7.14/rhino-1.7.14.jar
* https://repo1.maven.org/maven2/org/mozilla/rhino-engine/1.7.14/rhino-engine-1.7.14.jar

**Comment by blza** [comment]
Ok, so the only one that worked for me is simple
```sql
{% var: num=3 %}
select {{ script: ${num} - 2 }} one
```
but I had to disable `enable variables` in connection settings, as by default `${num}` is treated by dbeaver as a variable that needs substitution.

The following one 
```sql
select {{db.ch-play: select * from numbers(1, 10) }};
```
gives me error:
`SQL Error [HY000]: Failed to load configuration`

and the next one
```sql
select {{ script: [1,2,3,4] }} num_col;
```
also gives me error:
`SQL Error [12060] [HY000]: [Vertica][JDBC](12060) ResultSet object has been closed.`

I will try to investigate the later further as it might be on the Vertica side, like it's closing the result set jdbcx driver is expecting to be open or something, but with ch-play I'm stuck.

**Comment by zhicwu** [comment]
> SQL Error [HY000]: Failed to load configuration

To access named database connection `db.ch-play`, you need to declare the connection in `~/.jdbcx/db/ch-play.properties`, please refer to [this](https://github.com/jdbcx/jdbcx/blob/main/docker/app/.jdbcx/db/ch-play.propertes) used in docker image. Of course you need to download [ClickHouse JDBC driver](https://repo1.maven.org/maven2/com/clickhouse/clickhouse-jdbc/0.4.6/clickhouse-jdbc-0.4.6-http.jar) and put it in your `jdbcx.custom.classpath` directory.

> SQL Error [12060] [HY000]: [Vertica][JDBC] (12060) ResultSet object has been closed.

I'm not sure about this one, could be a bug in JDBCX. I tried to reproduce the issue using docker image `vertica/vertica-ce` with no luck. Will try again tomorrow.

**Comment by blza** [comment]
Got it about named connections and ClickHouse, tried the following with two instances of Vertica 11:
```sql
select {{ script: [1] }} num_col;
```
this one works, so only on the second iteration of say `select {{ script: [1,2] }} num_col;` there's the error `ResultSet object has been closed.`

**Comment by blza** [comment]
Thank you, now your examples do work!

Will there be an option to run something like this
```sql
select 1 
from {{ script: ["SAE","SAEBY"] }}.last_tracking_data
limit 10;
```
resulting in two consecutive selects (not combining results in any way, but getting every select say in different DBeaver results tab)?

**Comment by zhicwu** [comment]
Yes, I had an idea of adding another extension similar to `var` to better manage runtime behaviors. For examples:
```sql
-- only consider the outermost `runtime`
{% runtime: fetchSize=5000, maxRows=10000, queryPreference=MultiResultSet %}
select {{ script: [1,2] }} -- same as `select 1; select 2`

{% runtime: fetchSize=5000, maxRows=10000, queryPreference=Union %}
select {{ script: [1,2] }} -- same as `select 1 union all select 2`
```

I'm currently adding support for more data formats. Will get back to this in a few days. I'll also create a separate issue to list items I'm going to add in the near future.

**Comment by blza** [comment]
Did you implement it?
